book: To-Do 3 chapter (#1080)
gtk-rs/gtk4-rs: book/src/todo_3.md

# Let To-Do App Follow GNOME's HIG

> Within this chapter we will adapt our To-Do app so that it follows GNOME's [HIG](https://developer.gnome.org/hig/).

Author: It's already mentioned in the chapter before :)

Let's start by installing Libadwaita and adding the `libadwaita` crate to our dependencies as explained in the [former chapter](libadwaita.html).

The most simple way to take advantage of Libadwaita is by replacing [`gtk::Application`](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.Application.html) with [`adw::Application`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.Application.html).

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/5/main.rs">listings/todo/5/main.rs</a>

```rust,no_run,noplayground

> {{#rustdoc_include ../listings/todo/5/main.rs:main}}

Author: It's quite easy to add.

Feel free to point out similar things in future chapters or already existing ones.

But before we get to the interesting stuff, we make our lives easier in the future by replacing all occurrences of `gtk::prelude` and `gtk::subclass::prelude` with [`adw::prelude`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/prelude/index.html) and [`adw::subclass::prelude`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/subclass/prelude/index.html).
This works because the `adw` preludes re-export the corresponding `gtk` preludes plus a couple of Libadwaita specific traits.

In the remainder of this chapter we are going to follow a couple of patterns of GNOME's HIG.
Let's start by adapting the header bar to be [HIG compliant](https://developer.gnome.org/hig/patterns/containers/header-bars.html).
We do that by replacing [`gtk::ApplicationWindow`](../docs/gtk4/struct.ApplicationWindow.html) with [`adw::ApplicationWindow`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.ApplicationWindow.html).
We also use [`adw::HeaderBar`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.HeaderBar.html) as title bar.

> Finally, we add [tooltips](https://developer.gnome.org/hig/patterns/feedback/tooltips.html).

Author: The declared goal is to make the to-do app HIG compliant and that one requires tooltips for buttons in the header bar.

The second pattern we are going to follow are [boxed lists](https://developer.gnome.org/hig/patterns/containers/boxed-lists.html).

> The HIG does not require us to use them and there's a good reason for that: they are incompatible with recycling lists.

Reviewer: ```suggestion
The HIG does not require us to use this style and there's a good reason for that: it is incompatible with the recycling list.
```

This means they cannot be used with [list views](https://developer.gnome.org/hig/patterns/containers/list-column-views.html) and are therefore only appropriate for relatively small lists.

> > Try to add tasks programmatically and see how many of them you have to add until the UI noticeably slows down.

Reviewer: We're talking more than a thousand rows until it slows down I think

Author: 10 000 even on my machine.

> Determine for yourself if you think that is a reasonable number or if we should have rather stuck with list views.

We can use boxed lists by using [`gtk::ListBox`](../docs/gtk4/struct.ListBox.html) instead of [`gtk::ListView`](../docs/gtk4/struct.ListView.html).

> Afterwards, we match with the [`boxed-list`](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/main/boxed-lists.html) style class that is provided by Libadwaita.

Reviewer: ```suggestion
We will also add the [`boxed-list`](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/main/boxed-lists.html) style class provided by Libadwaita.
```

> Let's implement the discussed changes in the `window.ui` file.

Reviewer: ```suggestion
Let's implement all these changes in the `window.ui` file.
```

You can find the relevant subset of the diff below.

> To see the complete file, just click on the link after "Filename:".

Reviewer: Is this needed? Like, it's chapter 15 already :)

Author: Fair, but I only mention it once in the beginning of the book and this is the first time that I show an incomplete diff

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/6/resources/window.ui">listings/todo/6/resources/window.ui</a>

```diff
-  <template class="TodoWindow" parent="GtkApplicationWindow">
+  <template class="TodoWindow" parent="AdwApplicationWindow">
     <property name="width-request">360</property>
     <property name="title" translatable="yes">To-Do</property>
+    <property name="content">
       <object class="GtkBox">
         <property name="orientation">vertical</property>
+        <property name="hexpand">True</property>
         <child>
+          <object class="AdwHeaderBar">
+            <child type="end">
+              <object class="GtkMenuButton">
+                <property name="icon-name">open-menu-symbolic</property>
+                <property name="menu-model">main-menu</property>
+              </object>

> +            </child>

Reviewer: This diff doesn't look right. The tags on the added lines don't add up.

+                        <property name="margin-end">12</property>
+                        <child>
+                          <object class="GtkEntry" id="entry">
+                            <property name="placeholder-text" translatable="yes">Enter a Task…</property>
+                            <property name="secondary-icon-name">list-add-symbolic</property>
+                          </object>
+                        </child>
+                        <child>
+                          <object class="GtkListBox" id="tasks_list">
+                            <property name="visible">False</property>
+                            <property name="selection-mode">none</property>
+                            <style>
+                              <class name="boxed-list" />
+                            </style>
+                          </object>
+                        </child>
+                      </object>
+                    </property>
+                  </object>
+                </property>
+              </object>
+            </property>
           </object>
         </child>
       </object>
+    </property>
   </template>
 </interface>
```

> We replaced the `gtk::ApplicationWindow` with `adw::ApplicationWindow` and added a `adw::HeaderBar` to it.

Reviewer: ```suggestion
We've replaced the `gtk::ApplicationWindow` with `adw::ApplicationWindow` and added an `adw::HeaderBar` to it.
```

> In order to follow the boxed list pattern, we started using [`gtk::ListBox`](../docs/gtk4/struct.ListBox.html), set its property "selection-mode" to "none" and let it match with the `boxed-list` style class. 

Reviewer: ```suggestion
In order to follow the boxed list pattern, we switched to [`gtk::ListBox`](../docs/gtk4/struct.ListBox.html), set its property "selection-mode" to "none" and added the `boxed-list` style class. 
```

Let's continue with `window/imp.rs`.
The member variable `tasks_list` now describes a `ListBox` rather than a `ListView`.

> We also change the `ParentType` from `gtk::ApplicationWindow` to `adw::ApplicationWindow`.

Reviewer: ```suggestion
We also need to change the `ParentType` from `gtk::ApplicationWindow` to `adw::ApplicationWindow`.
```

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/6/window/imp.rs">listings/todo/6/window/imp.rs</a>

```rust,no_run,noplayground
{{#rustdoc_include ../listings/todo/6/window/imp.rs:window}}
```

We don't override any function of `adw::ApplicationWindow`, but we still have to add the empty `impl`.

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/6/window/imp.rs">listings/todo/6/window/imp.rs</a>

```rust,no_run,noplayground
{{#rustdoc_include ../listings/todo/6/window/imp.rs:AdwApplicationWindowImpl}}
```

We now move on to `window/mod.rs`.
First, we add `adw::ApplicationWindow` to our list of derived classes.

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/6/window/mod.rs">listings/todo/6/window/mod.rs</a>

```rust,no_run,noplayground
{{#rustdoc_include ../listings/todo/6/window/mod.rs:glib_wrapper}}
```

`ListBox` supports models just fine, but without any widget recycling we don't need factories anymore.
`setup_factory` can therefore be safely deleted.
To setup the `ListBox`, we call `bind_model` in `setup_tasks`.

> There we specify the model, as well as a closure describing how to transform the given GObject into a widget the list box can display. 

Reviewer: ```suggestion
There we specify the model, as well as a closure describing how to transform the given `GObject` into a widget the list box can display. 
```

Author: Leaving as is to stay consistent with other chapters

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/6/window/mod.rs">listings/todo/6/window/mod.rs</a>

```rust,no_run,noplayground
{{#rustdoc_include ../listings/todo/6/window/mod.rs:bind_model}}
```

We still have to specify the `create_task_row` method.
Here, we create an [`adw::ActionRow`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.ActionRow.html) with a [`gtk::CheckButton`](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.CheckButton.html) as activatable widget.

> Without recycling, a GObject will always belong to the same widget.

Reviewer: ```suggestion
Without recycling, a `GObject` will always belong to the same widget.
```

Author: Same here

That means we can just bind their properties without having to worry about unbinding them later on.

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/6/window/mod.rs">listings/todo/6/window/mod.rs</a>

```rust,no_run,noplayground
{{#rustdoc_include ../listings/todo/6/window/mod.rs:create_task_row}}
```

When using boxed lists, you also have to take care to hide the `ListBox` when there is no task present.

Filename: <a class=file-link href="https://github.com/gtk-rs/gtk4-rs/blob/master/book/listings/todo/6/window/mod.rs">listings/todo/6/window/mod.rs</a>

```rust,no_run,noplayground
{{#rustdoc_include ../listings/todo/6/window/mod.rs:connect_items_changed}}
```

> Finally, we specify the method `set_task_list_visible`.

Reviewer: ```suggestion
Finally, we define the `set_task_list_visible` method.
```
